_posts/2020-08-24-native-loader-net5.md

```diff
@@ -0,0 +1,38 @@
+
+In place of DllMap, users are encouraged to utilize the [NativeLibrary](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.nativelibrary?view=netcore-3.1) resolution APIs, which are set in managed code, and the [runtime hosting properties](https://docs.microsoft.com/en-us/dotnet/core/tutorials/netcore-hosting#step-3---prepare-runtime-properties), which are set by embedders with the `monovm_initialize` function. If these are insufficient for your use case, please [tell us about it](https://github.com/dotnet/runtime/issues/37213)! We're always looking to improve our interop functionality, and in particular with .NET 6 will be evaluating `NativeLibrary`, so please speak up.
+
+A more subtle, yet no less impactful change is that native library loading now defaults to `RTLD_LOCAL` to be consistent with CoreCLR and Windows, as opposed to our historical behavior of `RTLD_GLOBAL`. This change prevents symbol collision, and will both break and enable various scenarios/libraries. What this means in practice is that on Unix-like platforms, libraries are no longer loaded into a single global namespace and when looking up symbols, the library must be correctly specified.
+
+Historically, in Mono on Linux it was possible to load library `foo` containing symbol `bar`, and then invoke `bar` with a P/Invoke like so: 
+```
+[DllImport("asdf")]
+public static extern int bar();
+```
+
+This will no longer work. For that P/Invoke to function correctly, the attribute would need to use the correct library name: `[DllImport("foo")]`. A lot of code in the wild that was using incorrect library names will need to be updated. However, this means that when loading two libraries containing the same symbol name, there is no longer a conflict.
+
+There have been some embedding API changes as part of this. `MONO_DL_MASK` is no longer a full mask, as `MONO_DL_GLOBAL` has been introduced to specify `RTLD_GLOBAL`. If both `MONO_DL_LOCAL` and `MONO_DL_GLOBAL`, are set, Mono will use local. See mono/utils/mono-dl-fallback.h for more info.
+
+This also means that dynamically linking libmonosgen and attempting to resolve Mono symbols from `dlopen(NULL, ...)` will no longer work. `__Internal` has been preserved as a Mono-specific extension, but its meaning has been expanded. When P/Invoking into `__Internal`, the runtime will check both `dlopen(NULL)` and the runtime library in the case that they differ, so that users attempting to call Mono APIs with `__Internal` will not have those calls break.
+
+Mono now supports the `DefaultDllImportSearchPathsAttribute` attribute! In particular, passing `DllImportSearchPath.AssemblyDirectory` is now required to have the loader search the executing assembly's directory for native libraries.
```

Comment:
This paragraph (and the next couple) are separate changes, right?

It might be better to give this article some more structure (ie either add some headings, or make the whole thing a big `<ul>`)

---

Or maybe, since the other points are short, add the summary list at the beginning that just says "changed default from RTLD_GLOBAL to RTLD_LOCAL, see below for ramifications and caveats"  for the big change and add the smalelr ones just to the summary.

Reply:
I like the latter option, will update.

Reply:
@lambdageek took another swing at this, let me know if you think this is clearer.